### ISSUE TITLE
Provisioning of secondary files doesn't seem to rename properly

### ISSUE THREAD

**Issue by denis-yuen** [ga4gh/dockstore#1149]
### Desired behaviour

See https://gitter.im/ga4gh/dockstore?at=5a79c94b51915d962afef05c
It appears as though provisioning out to a final location renames the base file, but not the secondary files properly. 

Snippet of relevant json parameters
```
 "out_cram": {
    "path": "/tmp/mapped.cram",
    "class": "File",
    "secondaryFiles": [
      ".crai",
      ".bas",
      ".md5",
      ".met",
      ".maptime"
    ]
  }
```

Exposed by @keiranmraine

**Comment by keiranmraine** [comment]
@denis-yuen what's the chance of a short turnaround hotfix release for this?  Or is 1.4.0 imminent?

Happy to test.

**Comment by denis-yuen** [comment]
We'll do this as a hotfix since 1.4.0 is still more than a month away. 
I think this will be a couple days off though, due to grant-related activities, but I'll see if we can squeeze it in.

**Comment by keiranmraine** [comment]
Thanks, that's fits well with my timeline.  Just doing quite a large update of tools and want to use the cleanest approach for the json.

**Comment by keiranmraine** [comment]
I've been able to do some testing of the secondary file element of this and have found that although it is now renaming it doesn't seem to be matching the behaviour of input provisioning.

For an input defining a secondary file _adds_ the defined extension to the original and pulls that file.  So a secondary file of `.bas` on a primary of `test.bam` would also pull `test.bam.bas`.

On secondary files for output it's not doing this, it's removing the extension and replacing it:

```cwl
  "out_bam": {
    "path": "/tmp/mapped.bam",
    "class": "File",
    "secondaryFiles": [
      ".bai",
      ".bas",
      ".md5",
      ".met",
      ".maptime"
    ]
  }
```

Results in the following:

```
Provisioning your output files to their final destinations
Registering: #out_bam to provision from /.../test.bam to : /tmp/mapped.bam
Registering: #out_bam to provision from /.../test.bam.bai to : /tmp/mapped.bai
Registering: #out_bam to provision from /.../test.bam.bas to : /tmp/mapped.bas
Registering: #out_bam to provision from /.../test.bam.md5 to : /tmp/mapped.md5
Registering: #out_bam to provision from /.../test.bam.met to : /tmp/mapped.met
Registering: #out_bam to provision from /.../test.bam.maptime to : /tmp/mappedtime
```

I believe removal of the final component before adding the new one is only expected if the secondaryFile definition starts with `^`.


Additionally the extension seems to be limited to 4 characters resulting in mangling of the last entry, `.maptime` into `mappedtime` instead of `mapped.bam.maptime` (as `.bam` should be retained regardless of this bug).

Was this an intentional decision?  If so it's going to cause problems for CRAM and related indexes as the nomenclature for extensions is `.cram` and `.crai`.

**Comment by denis-yuen** [comment]
Looks like our testing on this is insufficient, thanks for the heads-up!

**Comment by keiranmraine** [comment]
I'm not sure what's going on here, possible that my json file is incorrect I guess.

The new 1.3.5 version doesn't seem to have changed anything.

CWL for output files is:

https://github.com/cancerit/dockstore-cgpmap/blob/5ca8160e240a9a25ee68c13ef87e36b3833a8a88/cwls/cgpmap-bamOut.cwl#L88-L98

Json for outputs is:

https://github.com/cancerit/dockstore-cgpmap/blob/5ca8160e240a9a25ee68c13ef87e36b3833a8a88/examples/cgpmap/bamOut/fastq_gz_input.json#L35-L45

Checked the new version is in use:

```
$ dockstore --version
Dockstore version 1.3.5
The latest stable version is 1.3.3
You are currently on the latest unstable version. If you wish to upgrade to the latest stable version, please use the following command:
   dockstore --upgrade-stable
```

Lanuched with:

```
dockstore tool launch --local-entry dockstore-cgpmap-feature-mmqc_metaRg/cwls/cgpmap-bamOut.cwl --json no_sec.json
```

Cwltool provisions output as expected:

```
cwltool stdout:
	{
	    "out_bam": {
	        "checksum": "sha1$f480fb2110298245c2647ef9bad0f3fafe1c273b", 
	        "basename": "test.bam", 
	        "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam", 
	        "secondaryFiles": [
	            {
	                "checksum": "sha1$5e4f9ffd56470bddf19adf3ddcff6e088185cf66", 
	                "basename": "test.bam.bai", 
	                "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bai", 
	                "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bai", 
	                "class": "File", 
	                "size": 951528
	            }, 
	            {
	                "checksum": "sha1$1a70259925f883e4a4856b225339e943c535996d", 
	                "basename": "test.bam.bas", 
	                "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bas", 
	                "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bas", 
	                "class": "File", 
	                "size": 660
	            }, 
	            {
	                "checksum": "sha1$bd48b4345e920e5ce36fa50ab7068a4889ebe9fc", 
	                "basename": "test.bam.md5", 
	                "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.md5", 
	                "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.md5", 
	                "class": "File", 
	                "size": 32
	            }, 
	            {
	                "checksum": "sha1$68cbd8e9cdbacaaed9148aed871a2a71162b0bb4", 
	                "basename": "test.bam.met", 
	                "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.met", 
	                "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.met", 
	                "class": "File", 
	                "size": 1483
	            }, 
	            {
	                "checksum": "sha1$7d46650b68a02a4e82fd9cc103d42dc7b9b0c9a0", 
	                "basename": "test.bam.maptime", 
	                "location": "file:///home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.maptime", 
	                "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.maptime", 
	                "class": "File", 
	                "size": 396
	            }
	        ], 
	        "path": "/home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam", 
	        "class": "File", 
	        "size": 6600215
	    }
	}
```

But Dockstore then does:

```
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam to : /tmp/mapped.bam
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bai to : /tmp/mapped.bai
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.bas to : /tmp/mapped.bas
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.md5 to : /tmp/mapped.md5
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.met to : /tmp/mapped.met
Registering: #out_bam to provision from /home/ubuntu/datastore/launcher-d54b4f1c-2f40-4876-88f4-12b071248f3a/outputs/test.bam.maptime to : /tmp/mappedtime
```

To be completely clear I expect the final block to be:

```
Registering: #out_bam to provision from /.../test.bam to : /tmp/mapped.bam
Registering: #out_bam to provision from /.../test.bam.bai to : /tmp/mapped.bam.bai
Registering: #out_bam to provision from /.../test.bam.bas to : /tmp/mapped.bam.bas
Registering: #out_bam to provision from /.../test.bam.md5 to : /tmp/mapped.bam.md5
Registering: #out_bam to provision from /.../test.bam.met to : /tmp/mapped.bam.met
Registering: #out_bam to provision from /.../test.bam.maptime to : /tmp/mapped.bam.maptime
```

**Comment by denis-yuen** [comment]
@keiranmraine  Apologies, looks like a gremlin in the release process. I've re-released 1.3.6 and you should see the following. 
```
$ dockstore --version
Downloading Dockstore to /home/dyuen/.dockstore/self-installs/dockstore-client-1.3.6.jar now...
file:////home/dyuen/.m2/repository/io/dockstore/dockstore-client/1.3.6/dockstore-client-1.3.6-shaded.jar
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 42.5M  100 42.5M    0     0  1143M      0 --:--:-- --:--:-- --:--:-- 1143M
Dockstore version 1.3.6
The latest stable version is 1.3.3
You are currently on the latest unstable version. If you wish to upgrade to the latest stable version, please use the following command:
   dockstore --upgrade-stable
dyuen@odl-dyuen2:~/dockstore_tools/dockstore-cgpmap$ dockstore tool launch --local-entry cwls/cgpmap-bamOut.cwl --json examples/cgpmap/bamOut/fastq_gz_input.json 
Creating directories for run of Dockstore launcher at: /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore//launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9
Provisioning your input files to your local machine
...
	Workflow end: Thu Feb 22 15:07:13 UTC 2018
	[job cgpmap-bamOut.cwl] completed success
	Final process status is success

Saving copy of cwltool stdout to: /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/cwltool.stdout.txt
Saving copy of cwltool stderr to: /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/cwltool.stderr.txt

Provisioning your output files to their final destinations
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam to : /tmp/mapped.bam
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.bai to : /tmp/mapped.bam.bai
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.bas to : /tmp/mapped.bam.bas
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.md5 to : /tmp/mapped.bam.md5
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.met to : /tmp/mapped.bam.met
Registering: #out_bam to provision from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.maptime to : /tmp/mapped.bam.maptime
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam to /tmp/mapped.bam
[##################################################] 100%
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.bai to /tmp/mapped.bam.bai
[##################################################] 100%
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.bas to /tmp/mapped.bam.bas
[##################################################] 100%
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.md5 to /tmp/mapped.bam.md5
[##################################################] 100%
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.met to /tmp/mapped.bam.met
[##################################################] 100%
Provisioning from /home/dyuen/dockstore_tools/dockstore-cgpmap/datastore/launcher-c1b5566f-3562-4c6a-9d54-a6fd8d2042d9/outputs/test.bam.maptime to /tmp/mapped.bam.maptime
[##################################################] 100%
```

**Comment by garyluu** [comment]
Looks fine to me:

```
Provisioning your output files to their final destinations
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam to : /tmp/mapped.bam
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.bai to : /tmp/mapped.bam.bai
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.bas to : /tmp/mapped.bam.bas
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.md5 to : /tmp/mapped.bam.md5
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.met to : /tmp/mapped.bam.met
Registering: #out_bam to provision from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.maptime to : /tmp/mapped.bam.maptime
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam to /tmp/mapped.bam
[##################################################] 100%
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.bai to /tmp/mapped.bam.bai
[##################################################] 100%
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.bas to /tmp/mapped.bam.bas
[##################################################] 100%
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.md5 to /tmp/mapped.bam.md5
[##################################################] 100%
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.met to /tmp/mapped.bam.met
[##################################################] 100%
Provisioning from /home/gluu/dockstore-cgpmap/datastore/launcher-1e3c34f8-89a6-4cf3-87b7-f9709ced81bb/outputs/test.bam.maptime to /tmp/mapped.bam.maptime
[##################################################] 100%
```
